bindings/go/src/fdb/range.go

```diff
@@ -170,119 +181,105 @@ func (rr RangeResult) GetSliceOrPanic() []KeyValue {
 func (rr RangeResult) Iterator() *RangeIterator {
 	return &RangeIterator{
 		t:         rr.t,
-		f:         rr.f,
 		sr:        rr.sr,
 		options:   rr.options,
 		iteration: 1,
 		snapshot:  rr.snapshot,
+		more:      true,
 	}
 }
 
 // RangeIterator returns the key-value pairs in the database (as KeyValue
 // objects) satisfying the range specified in a range read. RangeIterator is
 // constructed with the (RangeResult).Iterator method.
 //
-// You must call Advance and get a true result prior to calling Get or MustGet.
+// You must call NextBatch and get a true result.
 //
 // RangeIterator should not be copied or used concurrently from multiple
 // goroutines, but multiple RangeIterators may be constructed from a single
 // RangeResult and used concurrently. RangeIterator should not be returned from
 // a transactional function passed to the Transact method of a Transactor.
 type RangeIterator struct {
-	t         *transaction
-	f         *futureKeyValueArray
-	sr        SelectorRange
-	options   RangeOptions
-	iteration int
-	done      bool
-	more      bool
-	kvs       []KeyValue
-	index     int
-	err       error
-	snapshot  bool
+	t               *transaction
+	sr              SelectorRange
+	options         RangeOptions
+	iteration       int
+	more            bool
+	lastBatchLen    int
+	snapshot        bool
+	prefetchedBatch *futureKeyValueArray
+	prefetchOptions RangeOptions
 }
 
-// Advance attempts to advance the iterator to the next key-value pair. Advance
-// returns true if there are more key-value pairs satisfying the range, or false
-// if the range has been exhausted. You must call this before every call to Get
-// or MustGet.
-func (ri *RangeIterator) Advance() bool {
-	if ri.done {
-		return false
+// NextBatch advances the iterator to the next key-value pairs batch.
+// It returns a slice of key-value pairs satisfying the range, an empty slice
+// if the range has been exhausted, or an error if any of the asynchronous
+// operations associated with this result did not successfully complete.
+// If 'prefetch' is true, future for the next batch will be created on return and
+// released when the RangeIterator is closed.
+func (ri *RangeIterator) NextBatch(prefetch bool) ([]KeyValue, error) {
+	if !ri.more {
+		// iterator is done
+		return nil, nil
 	}
 
-	if ri.f == nil {
-		return true
+	var f *futureKeyValueArray
+	if ri.prefetchedBatch != nil {
+		// update state for the prefetched read batch
+		ri.options = ri.prefetchOptions
+		ri.iteration++
+
+		f = ri.prefetchedBatch
+		ri.prefetchedBatch = nil
+	} else {
+		f = ri.fetchNextBatch()
 	}
 
-	ri.kvs, ri.more, ri.err = ri.f.Get()
-	ri.index = 0
-	ri.f = nil
+	var err error
+	var kvs []KeyValue
+	kvs, ri.more, err = f.Get()
+	ri.lastBatchLen = len(kvs)
+	f.Close()
+	if err != nil {
+		// cannot re-use iterator after an error
+		ri.more = false
+		return nil, err
+	}
+	// consider reading done if limit was reached
+	ri.more = ri.more || ri.lastBatchLen == ri.options.Limit
```

Comment:
```suggestion
	ri.more = ri.more && (ri.options.Limit == 0 || ri.lastBatchLen < ri.options.Limit)
```

I believe the current code has a bug. We want to read "more" only if we haven't exceeded the limit yet (assuming that the limit is not `0`). In the current code if `ri.more` would be `true` we would continue even if we hit the limit and run another `doGetRange` with the limit being set to `0` in this case (because of the `ri.prefetchOptions.Limit -= ri.lastBatchLen`).

Reply:
I have rewritten it using an `if` branch, hopefully it helps with clarity?